File: Day-5/MaxManhattan.py
def max_manhattan_distance(s: str, k: int) -> int:
    # Count the number of each direction
    N = S = E = W = 0
    max_distance = 0

    for move in s:
        if move == 'N':
            N += 1
        elif move == 'S':
            S += 1
        elif move == 'E':
            E += 1
        elif move == 'W':
            W += 1

        # Greedy step: use up to k changes to reduce opposing directions

        # Vertical direction optimization
        min_vert = min(N, S)
        used_vert = min(k, min_vert)
        remaining = k - used_vert

        # Horizontal direction optimization
        min_horiz = min(E, W)
        used_horiz = min(remaining, min_horiz)

        # Compute current net distances after changes
        vertical_dist = abs(N - S) + 2 * used_vert
        horizontal_dist = abs(E - W) + 2 * used_horiz
        total_distance = vertical_dist + horizontal_dist

        # Update the maximum distance found
        max_distance = max(max_distance, total_distance)

    return max_distance

# Example test cases
print("Example 1:", max_manhattan_distance("NWSE", 1))  # Output: 3
print("Example 2:", max_manhattan_distance("NSWWEW", 3))  # Output: 6

File: Day-5/MaxManhattan.cpp
#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

// Function to compute maximum Manhattan distance
int maxManhattanDistance(const string& s, int k) {
    int N = 0, S = 0, E = 0, W = 0;
    int maxDist = 0;

    for (char move : s) {
        switch (move) {
            case 'N': N++; break;
            case 'S': S++; break;
            case 'E': E++; break;
            case 'W': W++; break;
        }

        // Greedy strategy: use k changes to reduce opposing directions

        // Vertical direction optimization
        int minVert = min(N, S);
        int usedVert = min(k, minVert);
        int remaining = k - usedVert;

        // Horizontal direction optimization
        int minHoriz = min(E, W);
        int usedHoriz = min(remaining, minHoriz);

        // Compute current net vertical and horizontal distances
        int verticalDist = abs(N - S) + 2 * usedVert;
        int horizontalDist = abs(E - W) + 2 * usedHoriz;
        int totalDist = verticalDist + horizontalDist;

        // Update the maximum distance found so far
        maxDist = max(maxDist, totalDist);
    }

    return maxDist;
}

// Main function with example test cases
int main() {
    string s1 = "NWSE";
    int k1 = 1;
    cout << "Example 1: " << maxManhattanDistance(s1, k1) << endl; // Output: 3

    string s2 = "NSWWEW";
    int k2 = 3;
    cout << "Example 2: " << maxManhattanDistance(s2, k2) << endl; // Output: 6

    return 0;
}